### ISSUE TITLE
Variable agnostic modelling using metadata representation and embedding

### ISSUE THREAD

**Issue by OpheliaMiralles** [ecmwf/anemoi-core#1411]
### Motivation
We want a variable-agnostic model that encodes heterogeneous regional datasets into a shared latent representation of the atmosphere and decodes the requested physical fields. The current code requires global and regional datasets to share the same variables when constructing a cutout dataset, and does not support resuming from a checkpoint with a different variable set. These restrictions tie weather services to the variables available in ERA5, IFS, or checkpoints released by other national centres.

The intersection of available variables can shrink as more domains are added. We want to work with the union of those variables, using all available information and, ideally, predicting requested fields through their metadata, including when they are absent from a particular source.

### Proposed representation

For each dataset, store:

```text
values:         [batch, nodes, fields]
field metadata: [batch, fields, metadata]
field IDs:      [batch, fields]
```

Metadata should be stored once per field, ideally not once per grid point. It can be broadcast temporarily while combining values with embedded metadata, processed in bounded node chunks, and pooled over fields before graph encoding.

### Metadata semantics

The learned field description should embed the metadata that defines the physical quantity:

```yaml
param: t
vertical_coordinate:
  type: pressure                 # height_above_ground | height_above_sea | pressure | model
  level: 850
  unit: hPa                      # not mandatory
temporal_operator: instantaneous  # instantaneous | accumulation | mean | maximum | minimum
temporal_window: null             # required for non-instantaneous fields
value_unit: K
```

Pressure and height should be represented as continuous physical coordinates in canonical units to be able to extrapolate to unseen levels. Model levels are not universal, so they require a provenance-specific vertical-coordinate definition or conversion to physical coordinates but could be dealt with in a second space.

### Algorithm

```text
native values + field metadata
    -> per-node field pooling
    -> encoder / processor / decoder
    -> output transform conditioned on requested field metadata
```

Pooling would remove the explicit variable axis before spatial encoding and give the graph components a fixed-width node representation.

### Embedding

Shared across nodes, the field metadata could combine:

- A categorical embedding of `param`.
- A continuous encoding of pressure or height, such as a positional encoding or learned projection.
- A categorical embedding of `vertical_coordinate.type`.
- Encodings of the temporal operator, window and units where relevant.

At each node, the field values, optionally passed through a value encoder, could be combined with these metadata embeddings. Pooling over the available fields could then use attention or another set aggregation method. The specific combination and pooling operations would remain configurable.

On the output side, a corresponding transform could combine the decoded node representation with embeddings of the requested fields. This would allow the output field set to differ from the input field set.

### Pipeline design

Since graph encoding and processing logic is shared with previous iterations of the code, we suggest isolating the embedding logic in an isolated configurable **`input_transform`** module that replaces the preprocessing of **`x_vars`** in **`_assemble_input`** in the current `enc_proc_dec` model:

```python
def _assemble_input(...) -> tuple[torch.Tensor, torch.Tensor, ShardSizes]:
    node_attributes_data = ...
    grid_shard_sizes = ...
    x_skip = ...
    x_vars = self.input_transform(x, node_attributes_data)  # configurable in config file
    return x_vars, x_skip, grid_shard_sizes
```

### Handling heterogeneous field sets

- **Missing fields and values:** omit absent fields from the active values and metadata. Variable dropout could expose the model to different field subsets during training while retaining a fixed-width pooled representation. Per-node validity masks could exclude missing observations from pooling within an otherwise available field. This is particularly relevant in the context multidomain training.

- **Normalisation and checkpoint compatibility:** normalisation statistics could be selected by physical field identity after conversion to canonical units. A persistent vocabulary, metadata definitions and normalisation state could be saved alongside the model, allowing a new dataset’s active fields to be mapped independently of channel order.

- **Residual/skip connections:** construct skip mappings from compatible input and output field identities. Shared fields could retain a residual connection; outputs without a corresponding input could be predicted directly from the latent representation. Field selection and variable dropout would update these mappings consistently.

- **Requested fields absent from a source:** query the output transform with the requested metadata, using relationships learned from training data that provide those targets. Predicting a physical variable never supervised during training would remain an open question.

### Design goals

- Accept different variables, levels and missing fields from each source.
- Store field metadata as `O(fields × metadata)`, while the spatial processor operates without a variable axis.
- Investigate decoding at continuous pressure and height levels, including levels not seen during training.

### Additional context

In the DE-376 project, differences in the variables available across regional datasets limit how much atmospheric information we can use. For example, we cannot currently build a stretched grid from datasets with different variable sets or resume training from a checkpoint with a different set of variables. We propose a metadata-based representation to overcome these restrictions, allowing the model to accommodate missing inputs and predict requested fields even when they are absent from a particular training source.

### Organisation

DE-376

### Additional notes

Would configurable input and output transforms be a suitable integration point? Feedback on the metadata structure and compatibility with existing normalisation and checkpoint handling would also be useful.

### Edge cases identified:
- `mslp` : should be `param` `p` and level `height_above_sea` vs `sp` should be `param` `p` and level `height_above_ground`

**Comment by einrone** [comment]
current vocabular based on the metadata from the zarr archives:
```
VariableSpecification(name='10u', param='u', vertical_coordinate=VerticalCoordinate(type='sfc', level=10, unit='m'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='10v', param='v', vertical_coordinate=VerticalCoordinate(type='sfc', level=10, unit='m'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='2t', param='t', vertical_coordinate=VerticalCoordinate(type='sfc', level=2, unit='m'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='cos_latitude', param='cos_latitude', vertical_coordinate=VerticalCoordinate(type='forcing', level=None, unit=None), temporal_operator=None, temporal_window=None)
VariableSpecification(name='cos_longitude', param='cos_longitude', vertical_coordinate=VerticalCoordinate(type='forcing', level=None, unit=None), temporal_operator=None, temporal_window=None)
VariableSpecification(name='msl', param='msl', vertical_coordinate=VerticalCoordinate(type='sfc', level=0, unit='m'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='q_500', param='q', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='q_850', param='q', vertical_coordinate=VerticalCoordinate(type='pl', level=850, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='sin_latitude', param='sin_latitude', vertical_coordinate=VerticalCoordinate(type='forcing', level=None, unit=None), temporal_operator=None, temporal_window=None)
VariableSpecification(name='sin_longitude', param='sin_longitude', vertical_coordinate=VerticalCoordinate(type='forcing', level=None, unit=None), temporal_operator=None, temporal_window=None)
VariableSpecification(name='t_500', param='t', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='t_850', param='t', vertical_coordinate=VerticalCoordinate(type='pl', level=850, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='tp', param='tp', vertical_coordinate=VerticalCoordinate(type='sfc', level=0, unit='m'), temporal_operator='accumulated', temporal_window=1800)
VariableSpecification(name='u_500', param='u', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='u_850', param='u', vertical_coordinate=VerticalCoordinate(type='pl', level=850, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='v_500', param='v', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='v_850', param='v', vertical_coordinate=VerticalCoordinate(type='pl', level=850, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='w_500', param='w', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
VariableSpecification(name='z_500', param='z', vertical_coordinate=VerticalCoordinate(type='pl', level=500, unit='hpa'), temporal_operator='instantenous', temporal_window=0)
``

**Comment by einrone** [comment]
vocabulary list updated. Forcings are not part of the dataclass as this will get embedded separately during the tokenization.
```
{'10u': VariableSpecification(name='10u', param='u', vertical_coordinate=VerticalCoordinate(type='heightaboveground', level=10.0, unit='m'), temporal_operator='instantaneous', temporal_window=None), '10v': VariableSpecification(name='10v', param='v', vertical_coordinate=VerticalCoordinate(type='heightaboveground', level=10.0, unit='m'), temporal_operator='instantaneous', temporal_window=None), '2t': VariableSpecification(name='2t', param='t', vertical_coordinate=VerticalCoordinate(type='heightaboveground', level=2.0, unit='m'), temporal_operator='instantaneous', temporal_window=None), 'msl': VariableSpecification(name='msl', param='msl', vertical_coordinate=VerticalCoordinate(type='surface', level=None, unit=None), temporal_operator='instantaneous', temporal_window=None), 'q_500': VariableSpecification(name='q_500', param='q', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'q_850': VariableSpecification(name='q_850', param='q', vertical_coordinate=VerticalCoordinate(type='pressure', level=850.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 't_500': VariableSpecification(name='t_500', param='t', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 't_850': VariableSpecification(name='t_850', param='t', vertical_coordinate=VerticalCoordinate(type='pressure', level=850.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'tp': VariableSpecification(name='tp', param='tp', vertical_coordinate=VerticalCoordinate(type='surface', level=None, unit=None), temporal_operator='accumulation', temporal_window=1800.0), 'u_500': VariableSpecification(name='u_500', param='u', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'u_850': VariableSpecification(name='u_850', param='u', vertical_coordinate=VerticalCoordinate(type='pressure', level=850.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'v_500': VariableSpecification(name='v_500', param='v', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'v_850': VariableSpecification(name='v_850', param='v', vertical_coordinate=VerticalCoordinate(type='pressure', level=850.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'w_500': VariableSpecification(name='w_500', param='w', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None), 'z_500': VariableSpecification(name='z_500', param='z', vertical_coordinate=VerticalCoordinate(type='pressure', level=500.0, unit='hPa'), temporal_operator='instantaneous', temporal_window=None)}
``

**Comment by OpheliaMiralles** [comment]
great! I don't see `mslp` and `sp` which should probably be `p` param for vertical level `height_above_sea` and `height_above_ground` with `level` 0 respectively, and I think `tp` needs to be `height_above_ground` 0 but otherwise perfect
